documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc

```diff
@@ -0,0 +1,250 @@
+* Prevent cache misses by preloading data
+* For JDBC Cache Store:
+** Use indexes on `id` column to prevent table scans
+** Use PRIMARY_KEY on `id` column
+** Configure batch-size, fetch-size, etc
+
+== Hints for program developers
+
+There are also several hints for developers which can be easily applied to the client application and will boost up the performance.
+
+=== Ignore return values
+
+When you're not interested in returning value of the `#put(k, v)` or `#remove(k)` method, use `Flag.IGNORE_RETURN_VALUES` flag as shown below:
+
+[source,java]
+.Using Flag.IGNORE_RETURN_VALUES
+----
+Cache noPreviousValueCache = cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES);
+noPreviousValueCache.put(k, v);
+----
+
+It is also possible to set this flag using ConfigurationBuilder
+
+[source,java]
+.Using ConfigurationBuilder settings
+----
+ConfigurationBuilder cb = new ConfigurationBuilder();
+cb.unsafe().unreliableReturnValues(true);
+----
+
+=== Use `Externalizer` for marshalling
+
+Infinispan uses JBoss Marshalling to transfer objects over the wire. The most efficient way to marshall user data is to provide an AdvancedExternalizer.
+This solutions prevents JBoss Marshalling from sending class name over the network and allows to save some bandwidth:
+
+[source,java]
+.User entity with Externalizer
+----
+import org.infinispan.marshall.AdvancedExternalizer;
+
+public class Book {
+
+   final String name;
+   final String author;
+
+   public Book(String name, String author) {
+      this.name = name;
+      this.author = author;
+   }
+
+   public static class BookExternalizer
+            implements AdvancedExternalizer<Book> {
+
+      @Override
+      public void writeObject(ObjectOutput output, Book book)
+            throws IOException {
+         output.writeObject(book.name);
+         output.writeObject(book.author);
+      }
+
+      @Override
+      public Person readObject(ObjectInput input)
+            throws IOException, ClassNotFoundException {
+         return new Person((String) input.readObject(), (String) input.readObject());
+      }
+
+      @Override
+      public Set<Class<? extends Book>> getTypeClasses() {
+         return Util.<Class<? extends Book>>asSet(Book.class);
+      }
+
+      @Override
+      public Integer getId() {
+         return 2345;
+      }
+   }
+}
+----
+
+The Externalizer must be registered in cache configuration. See configuration examples below:
+
+[source,java]
+.Adding Externalizer using XML
+----
+<cache-container>
+   <serialization>
+      <advanced-externalizer class="Book$BookExternalizer"/>
+   </serialization>
+</cache-container>
+----
+
+[source,java]
+.Adding Externalizer using Java
+----
+GlobalConfigurationBuilder builder = ...
+builder.serialization().addAdvancedExternalizer(new Book.BookExternalizer());
+----
+
+For more information, please refer to the http://infinispan.org/docs/{infinispanversion}/user_guide/user_guide.html#benefits_of_externalizers[User's Guide].
```

Comment:
Here you should use a relative link:
link:$$../user_guide/user_guide.html$$[User's Guide]